Store hunk assignments in uncommitted file CliID

jonathantanmy2 opened the pull request:
This is the first step towards unifying uncommitted file and uncommitted hunk CLI IDs, and at the same time preventing some double work (of the consumers of the CLI IDs recalculating the assignments given the filename and stack assignment).

There are some code annoyances where I would have liked to state in the type system that a Vec has at least one element (in other words, it has an infallible `first()`). Any ideas? I can think of a few:
- depend on a Vec1 crate
- make our own Vec1 crate in our codebase (this prevents adding a dependency, and we don't actually need that many methods as far as I can tell - `from_element`, `iter`, `into_iter`, `first`, maybe a `map_vec1` that maps straight to another Vec1 since we cannot use `iter` to do it without losing the guarantee that there's at least one element)
- expand it into `foo_first` and `foo_rest` fields whenever we need this guarantee

There are also other options like having something like `foos: Vec<Foo>; foo_first: &Foo` where `foo_first`'s lifetime is constrained to that of `foos`, but I think it gets complicated if we need to `into_iter` from `foos` to another data structure where we also want to enforce that there's at least one element.

I have encountered this problem in this PR, but am encountering it even more in another WIP change (getting rid of `resolve_uncommitted_file_or_unassigned`), so if we could come up with a solution, I would appreciate it.

jonathantanmy2 committed:
id: split into modules

The main file was getting complicated, so split it into different files.

jonathantanmy2 committed:
id: hunk assignments in uncommitted file `CliId`

This prevents double work, and prepares for future work that will unify
`CliId`s for uncommitted files and hunks.

There is some code in `rub` that generates every uncommitted file
`CliId` while parsing ranges, which will incur additional memory
copying, but this will hopefully go away when range parsing is
refactored to be in `IdMap` instead (in the future).

jonathantanmy2 committed:
id: cleanup `CliId` traits

`CliId` does not need `Hash`, so remove it. Before some recent
significant changes, `Ord` was more crucial because we needed to
sort IDs when the user provided an ambiguous entity. However, most
ambiguities have been removed, so we no longer need to sort. The only
ambiguities remaining are:
- branch vs branch (handled by switching `branch_name_to_cli_id` to a
  `BTreeMap` to make the order consistent)
- commit ID vs unassigned area (handled by checking commits first)

So `Ord` is no longer necessary, so remove it too. Sorting is necessary
for tests because we list all IDs there, so add that functionality to
the tests.

As for `PartialEq`, change it to only check what's needed.

All this prepares `CliId` for a subsequent commit in which we add
a field that cannot have these traits derived easily.

jonathantanmy2 committed:
rub: make `source` owned

In a subsequent commit, `CliId` will contain a more substantial data
structure, so make `source` owned to improve performance.

jonathantanmy2 committed:
id: remove resolve_uncommitted_file_or_unassigned

This also removes some complicated parts of the code, including
`UncommittedFile` needing to be indexable in a `BTreeSet` both by (stack
ID, path) and by CLI ID (as of this commit, it only needs to be stored
in a regular map). This also makes it possible to use a `NonEmpty`
in `CliId::UncommittedFile`, with associated code improvements in its
users, because we no longer need a fallback CLI ID if the uncommitted
file does not have an associated ID (this situation is now impossible).

This commit also eliminates some duplicate work in callers. Previously,
some callers obtained hunk assignments from elsewhere and reconciled
them with `IdMap` using `resolve_uncommitted_file_or_unassigned`.
These callers have been changed to use the data structures in `IdMap`
directly.

jonathantanmy2 committed:
id: ensure committed file info is sorted

Info about committed files will later go into a `BTreeSet` with specific
ordering requirements; the caller in `mod.rs` assumes the invariants
documented in `FileInfo`. In `from_workspace_commits_and_status`, sort
the info so that we maintain those invariants.

Byron committed:
refactor

- clarify fixture support and fully parameterize it to make calls clear.
- adjust expectations to match the changed CLIIds

Byron committed:
Merge pull request #11609 from gitbutlerapp/jt/cffix

id: ensure committed file info is sorted

Byron commented on crates/but/src/command/legacy/rub/mod.rs:
```diff
@@ -298,71 +310,33 @@ fn parse_range(ctx: &mut Context, id_map: &IdMap, source: &str) -> anyhow::Resul
     ))
 }
 fn get_all_files_in_display_order(ctx: &mut Context, id_map: &IdMap) -> anyhow::Result<Vec<CliId>> {
```
I didn't realise this before, but my guess here is that this will bite as soon as it can, unless I am getting this wrong in thinking that `but rub` is coupled with `but status` because of the range support where display order matters.

Byron committed:
refactor

- fixup an issue introduced by me earlier (but didn't wait for CI apparently)
- make it easier to run rustdoc with decent performance after they broke it

Byron commented on crates/but/src/id/file_info.rs:
```diff
@@ -0,0 +1,60 @@
+                .and_modify(|hunk_assignments| {
+                    hunk_assignments.push(assignment.clone());
+                })
+                .or_insert_with(|| NonEmpty::new(assignment.clone()));
```
It's documented why that isn't possible here.

Byron reviewed (approved):
Thanks a lot, `nonempty` is a great choice and I trust the authors.
Also, I'd hope we can use `NonEmpty` more in the codebase now, it's definitely nice to use and great to remove more expect/unwrap, making the code more readable and self-explanatory.
